proptest-state-machine/src/strategy.rs

```diff
@@ -0,0 +1,761 @@
+                ix_to_check = 0;
+            } else {
+                ix_to_check += 1;
+            }
+            // We're back to where we started, there nothing left to do
+            if ix_to_check == ix {
+                return false;
+            }
+        }
+    }
+
+    /// Check if the sequence of included transitions is acceptable by the
+    /// pre-conditions. When `ix` is not `None`, the transition at the given
+    /// index is taken from its current value.
+    fn check_acceptable(&self, ix: Option<usize>) -> bool {
+        let transitions = self.get_included_acceptable_transitions(ix);
+        let mut state = self.last_valid_initial_state.clone();
+        for transition in transitions.iter() {
+            let current_acceptable = (self.preconditions)(&state, transition);
+            if current_acceptable {
+                state = (self.next)(state, transition);
+            } else {
+                return false;
+            }
+        }
+        true
+    }
+
+    /// The currently included and acceptable transitions. When `ix` is not
+    /// `None`, the transition at this index is taken from its current value
+    /// which may not be acceptable by the pre-conditions, instead of its
+    /// acceptable value.
+    fn get_included_acceptable_transitions(
+        &self,
+        ix: Option<usize>,
+    ) -> Vec<Transition> {
+        self.acceptable_transitions
+            .iter()
+            .enumerate()
+            // Filter out deleted transitions
+            .filter(|&(this_ix, _)| self.included_transitions.test(this_ix))
+            // Map the indices to the values
+            .map(|(this_ix, (_, transition))| match ix {
+                Some(ix) if this_ix == ix => self.transitions[ix].current(),
+                _ => transition.clone(),
+            })
+            .collect()
+    }
+
+    /// Find if the initial state is still shrinkable or if any of the
+    /// simplifications and complications of the included transitions have not
+    /// yet been rejected.
+    fn can_simplify(&self) -> bool {
+        self.is_initial_state_shrinkable ||
+             // If there are some transitions whose shrinking has not yet been 
+             // rejected, we can try to shrink them further
+             !self
+                .acceptable_transitions
+                .iter()
+                .enumerate()
+                // Filter out deleted transitions
+                .filter(|&(ix, _)| self.included_transitions.test(ix))
+                .all(|(_, (state, _transition))| {
+                    matches!(state, SimplifyRejected | ComplicateRejected)
+                })
+    }
+
+    /// Find the next shrink transition. Loops back to the front of the list
+    /// when the end is reached.
+    fn next_shrink_transition(&self, current_ix: usize) -> Shrink {
+        if current_ix == self.max_ix {
+            // Either loop back to the start of the list...
+            Transition(0)
```

Comment:
Why is a rotation needed? After deleting from the end we start simplification at `self.shrink = Transition(0);` and step forward, do we not cover everything in one pass? Can we document why that is?

Reply:
I've added extra line here [bd6d95c](https://github.com/proptest-rs/proptest/pull/310/commits/bd6d95cc76c65b593da11d7135d057c64d90cb00), hope that's clearer